client/js/analysis/gdpr.js

```diff
@@ -67,14 +79,120 @@ export async function getStreamingData(data) {
+                        return entry;
+                      });
+
+                  resolve(collatedEntries);
                 });
           });
         },
         (error) => reject(error),
     );
   });
 }
+
+/** @typedef {Map<string, Map<string, number>>} ArtistTrackTimeMap */
+/** @typedef {{
+ * start: Date,
+ * totals: ArtistTrackTimeMap
+ * }} ArtistTrackTimeInterval */
+
+/**
+ * Calculates aggregate per-track listening times from a list of GDPR records.
+ *
+ * @param {GDPRRecord[]} data The list of GDPR records to aggregate.
+ * @returns {{totals: ArtistTrackTimeMap, intervals: ArtistTrackTimeInterval[]}}
+ * An object with two properties: `totals` and `intervals`. `totals` is a map
+ * whose keys are artists and whose values are maps from track names to duration
+ * listened. `intervals` is an array of interval objects. Each interval object
+ * has a `start` representing the start of that interval, and a `totals`,
+ * representing the durations listened to each song during that interval.
+ */
+export function collateStreamingData(data) {
+  const intervals = [];
+
+  // a map of maps: key is artist name, value is a map where key
+  // is track name and value is time listened to that song in total
+  /** @type {ArtistTrackTimeMap} */
+  const totals = new Map();
+
+  // each interval is a map of maps: key is artist name, value is a map where
+  // key is track name and value is time listened to that song during this
+  // interval
+  /** @type {ArtistTrackTimeMap} */
+  let interval = new Map();
+
+  let index = 0;
+  let currentRecord = data[index];
+
+  const INTERVAL_SIZE = 24 * 60 * 60 * 1000;
+  // round down to the start of the day of the first record
+  let intervalStart =
+    +currentRecord.endTime - (+currentRecord.endTime % INTERVAL_SIZE);
+
+  while (index < data.length) {
+    while (currentRecord &&
+        +currentRecord.endTime < intervalStart + INTERVAL_SIZE) {
+      const {artistName, trackName, msPlayed} = currentRecord;
+
+      // update time in interval
+      let intervalArtistTracks = interval.get(artistName);
+
+      if (!intervalArtistTracks) {
+        intervalArtistTracks = new Map();
+        interval.set(artistName, intervalArtistTracks);
+      }
+
+      let intervalTrackTime = intervalArtistTracks.get(trackName);
+
+      if (!intervalTrackTime) {
+        intervalTrackTime = 0;
+      }
+
+      intervalTrackTime += msPlayed;
+
+      intervalArtistTracks.set(trackName, intervalTrackTime);
+
+      // update total time
+      let totalArtistTracks = totals.get(artistName);
+
+      if (!totalArtistTracks) {
+        totalArtistTracks = new Map();
+        totals.set(artistName, totalArtistTracks);
+      }
+
+      let totalTrackTime = totalArtistTracks.get(trackName);
+
+      if (!totalTrackTime) {
+        totalTrackTime = 0;
+      }
+
+      totalTrackTime += msPlayed;
+
+      totalArtistTracks.set(trackName, totalTrackTime);
+
+      currentRecord = data[++index];
```

Comment:
nit: separate the array index access and the incrementation. These should be on separate lines to improve readability.